wallet/cardano-trezor-crypto-provider.js

```diff
@@ -0,0 +1,334 @@
+const TrezorConnect = require('trezor-connect')
+const BlockchainExplorer = require('./blockchain-explorer')
+const derivePublic = require('./helpers/derivePublic')
+const {packAddress, unpackAddress} = require('./address')
+
+const CardanoTrezorCryptoProvider = (CARDANOLITE_CONFIG, walletState) => {
+  const state = Object.assign(walletState, {
+    derivedXpubs: {},
+    rootHdPassphrase: null,
+    derivedAddresses: {},
+  })
+
+  const blockchainExplorer = BlockchainExplorer(CARDANOLITE_CONFIG, state)
+
+  function toBip44Path(derivationPath) {
+    // 44'/1815'
+    let resultPath = [2147483692, 2147485463]
+
+    if (derivationPath.length === 2) {
+      resultPath = resultPath.concat([derivationPath[0], 0, derivationPath[1]])
+    } else if (derivationPath.length === 1) {
+      resultPath.push(derivationPath[0])
+    }
+
+    return resultPath
+  }
+
+  async function getWalletId() {
+    return await deriveAddress([], 'hardened')
+  }
+
+  async function deriveAddresses(derivationPaths, derivationMode) {
+    const addresses = []
+
+    derivationPaths = derivationPaths.filter((path) => {
+      const cachedAddress = state.derivedAddresses[JSON.stringify(path)]
+      if (cachedAddress) {
+        addresses.push(cachedAddress.address)
+        return false
+      }
+
+      return true
+    })
+
+    if (!derivationPaths.length) {
+      return addresses
+    }
+
+    let newAddresses = []
+    if (derivationMode === 'hardened') {
+      newAddresses = await trezorDeriveAddresses(derivationPaths, true)
+    } else {
+      for (const derivationPath of derivationPaths) {
+        newAddresses.push(await deriveAddress(derivationPath, derivationMode))
+      }
+    }
+
+    return addresses.concat(newAddresses)
+  }
+
+  function fromBip44Path(path) {
+    switch (path.length) {
```

Comment:
this can be generalized like this:

```
if ([0,3,5].includes(path.length)) {
    return path.filter((elem, i) => [2,4].includes(i))
} else {
    throw Error(`Unsupported derivation path: ${path}`)
}
```

Reply:
Yes it can, but I am not sure if it is also more readable a easy to understand.

Reply:
ok I agree